apollo-router/src/plugins/telemetry/mod.rs

```diff
@@ -0,0 +1,522 @@
+            trace_config
+        }
+    }
+}
+
+fn parent_based(sampler: opentelemetry::sdk::trace::Sampler) -> opentelemetry::sdk::trace::Sampler {
+    opentelemetry::sdk::trace::Sampler::ParentBased(Box::new(sampler))
+}
+
+#[derive(Debug)]
+struct ReportingError;
+
+impl fmt::Display for ReportingError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "ReportingError")
+    }
+}
+
+impl std::error::Error for ReportingError {}
+
+trait TracingConfigurator {
+    fn apply(&self, builder: Builder, trace_config: &Trace) -> Result<Builder, BoxError>;
+}
+
+trait GenericWith<T>
+where
+    Self: Sized,
+{
+    fn with<B>(self, option: &Option<B>, apply: fn(Self, &B) -> Self) -> Self {
+        if let Some(option) = option {
+            return apply(self, option);
+        }
+        self
+    }
+    fn try_with<B>(
+        self,
+        option: &Option<B>,
+        apply: fn(Self, &B) -> Result<Self, BoxError>,
+    ) -> Result<Self, BoxError> {
+        if let Some(option) = option {
+            return apply(self, option);
+        }
+        Ok(self)
+    }
+}
+
+impl<T> GenericWith<T> for T where Self: Sized {}
+
+fn setup<T: TracingConfigurator>(
+    mut builder: Builder,
+    configurator: &Option<T>,
+    tracing_config: &Trace,
+) -> Result<Builder, BoxError> {
+    if let Some(config) = configurator {
+        builder = config.apply(builder, tracing_config)?;
+    }
+    Ok(builder)
+}
+
+fn apollo_key() -> Option<String> {
+    std::env::var("APOLLO_KEY").ok()
+}
+
+fn apollo_graph_reference() -> Option<String> {
+    std::env::var("APOLLO_GRAPH_REF").ok()
+}
+
+#[async_trait::async_trait]
+impl Plugin for Telemetry {
+    type Config = config::Conf;
+
+    async fn startup(&mut self) -> Result<(), BoxError> {
+        // Apollo config is special because we enable tracing if some env variables are present.
+        let apollo = self.config.apollo.get_or_insert_with(Default::default);
+        if apollo.apollo_key.is_none() {
+            apollo.apollo_key = apollo_key()
+        }
+        if apollo.apollo_graph_ref.is_none() {
+            apollo.apollo_graph_ref = apollo_graph_reference()
+        }
+
+        // If we have key and graph ref but no endpoint we start embedded spaceport
+        let (spaceport, shutdown_tx) = match apollo {
+            apollo::Config {
+                apollo_key: Some(_),
+                apollo_graph_ref: Some(_),
+                endpoint: None,
+            } => {
+                let (shutdown_tx, shutdown_rx) = futures::channel::oneshot::channel();
+                let report_spaceport = ReportSpaceport::new(
+                    "127.0.0.1:0".parse()?,
+                    Some(Box::pin(shutdown_rx.map(|_| ()))),
+                )
+                .await?;
+                // Now that the port is known update the config
+                apollo.endpoint = Some(Url::parse(&format!(
+                    "https://{}",
```

Comment:
spaceport is listening locally in https?

Reply:
It is I think: https://github.com/apollographql/router/blob/main/apollo-router/src/apollo_telemetry.rs#L53

@garypen How can this work? Is it just using an anonymous cert?

Reply:
Good question. We don't even have TLS enabled in our tonic crate so I don't see how https could work at all. I think that the tonic connector must be downgrading to http (silently). I just tried trawling through the source to confirm that, but couldn't find anything yet. We should change these "https" specifications to "http" for now until we figure out how we want to enable TLS/https for external spaceports. (We probably only want "http" for internal spaceports anyway.)